# Setting up a virtual environment

Creating a virtual environment is helpful to manage packages for different projects. If you need different versions of the same package for various projects loading them into the global environment could lead to breaking a project. With a virtual environment, only those packages are loaded for the project that are necessary.

#### 1. Create the environment with the name "project_challenge".

In [1]:
conda create -n project_challenge -y

Solving environment: ...working... done

## Package Plan ##

  environment location: C:\Users\jana\anaconda3\envs\project_challenge



Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done
#
# To activate this environment, use
#
#     $ conda activate project_challenge
#
# To deactivate an active environment, use
#
#     $ conda deactivate

Retrieving notices: ...working... done

Note: you may need to restart the kernel to use updated packages.


#### 2. Check if the creation was successful.

You can see that we now have the base environment as well as the environment "project_challenge" which we just created.

In [2]:
conda env list

# conda environments:
#
base                     C:\Users\jana\anaconda3
project_challenge     *  C:\Users\jana\anaconda3\envs\project_challenge


Note: you may need to restart the kernel to use updated packages.


#### 3. After creating the environment, we must activate it to use it.

In [3]:
conda activate project_challenge


Note: you may need to restart the kernel to use updated packages.


In [4]:
conda env list

# conda environments:
#
base                     C:\Users\jana\anaconda3
project_challenge     *  C:\Users\jana\anaconda3\envs\project_challenge


Note: you may need to restart the kernel to use updated packages.


#### 4. Now, we can install all required packages into the virtual environment.

In [5]:
pip install -r requirements.txt

  Using cached argon2_cffi-20.1.0-cp37-cp37m-win_amd64.whl (42 kB)
  Using cached async_generator-1.10-py3-none-any.whl (18 kB)
  Using cached attrs-20.3.0-py2.py3-none-any.whl (49 kB)
  Using cached bleach-3.3.0-py2.py3-none-any.whl (283 kB)
  Using cached cachetools-4.2.1-py3-none-any.whl (12 kB)
  Using cached cffi-1.14.5-cp37-cp37m-win_amd64.whl (178 kB)
  Using cached chardet-4.0.0-py2.py3-none-any.whl (178 kB)
  Using cached decorator-4.4.2-py2.py3-none-any.whl (9.2 kB)
  Using cached google_api_core-1.26.0-py2.py3-none-any.whl (92 kB)
  Using cached google_auth-1.26.1-py2.py3-none-any.whl (116 kB)
  Using cached google_cloud_bigquery-2.8.0-py2.py3-none-any.whl (212 kB)
  Using cached google_cloud_bigquery_storage-2.2.1-py2.py3-none-any.whl (140 kB)
  Using cached google_cloud_core-1.6.0-py2.py3-none-any.whl (28 kB)
  Using cached google_crc32c-1.1.2-cp37-cp37m-win_amd64.whl (34 kB)
  Using cached google_resumable_media-1.2.0-py2.py3-none-any.whl (75 kB)
  Using cached googleapis

Reason for being yanked: Name generation process created inappropriate id values
ERROR: Cannot uninstall 'terminado'. It is a distutils installed project and thus we cannot accurately determine which files belong to it which would lead to only a partial uninstall.


# Connecting Jupiter Notebook with Google Big Query API

To get access to the project created on google cloud, we need to tell Python where to find the credentials. 
This works like a key to get access to the project and to use bigquery in this Notebook

#### 1. We have to point at the location where the Credentials of the project are.

In [6]:
import os
os.environ['GOOGLE_APPLICATION_CREDENTIALS'] = "C:/Users/jana/Coding Challenge/learnings-code-challenges-bc3ef22c46f8.json"

# "Will it snow tomorrow?" - The time traveler asked
The following dataset contains climate information form over 9000 stations accross the world. The overall goal of these subtasks will be to predict whether it will snow tomorrow 13 years ago. So if today is 2022.02.15 then the weather we want to forecast is for the date 2009.02.16. You are suppsed to solve the tasks using Big Query, which can be used in the Jupyter Notebook like it is shown in the following cell. For further information and how to used BigQuery in Jupyter Notebook refer to the Google Docs. 

The goal of this test is, to test your coding knowledge in Python, BigQuery and Pandas as well as your understanding of Data Science. If you get stuck at the first part, you can use the replacement data provided in the second part

In [7]:
%reload_ext google.cloud.bigquery

#### 1. First, lets take a look on the first 20 rows of the table of the public data set of bigquery. 

The data contains weather information from different weather stations on different dates. One day and weather station is on observation. 

Information in the data set:
Besides basic information like the avarage temperature there is also information if the station reported fog, snow, hail, thrunder or tornado.    

In [8]:
%%bigquery 
SELECT
*,
FROM `bigquery-public-data.samples.gsod`
LIMIT 20 


Downloading: 100%|██████████| 20/20 [00:01<00:00, 11.15rows/s]


station_number  wban_number  year  month  day  mean_temp  \
0            37950        99999  1929     12    5  48.700001   
1            31590        99999  1929     10    2  50.200001   
2            38640        99999  1929     10    1  55.000000   
3            38110        99999  1929     12    8  48.700001   
4            30050        99999  1929     12   27  42.500000   
5            38110        99999  1929     10    4  53.299999   
6            38560        99999  1929     10   20  52.000000   
7            30910        99999  1929     10    4  43.799999   
8            38940        99999  1929     11   11  50.799999   
9            38640        99999  1929     11    1  44.700001   
10           38640        99999  1929     11   12  46.500000   
11           38640        99999  1929     11   19  51.799999   
12           39730        99999  1929     11   28  48.500000   
13           33790        99999  1929     12   17  33.299999   
14           38640        99999  1929     12   24  46.299999   
15           38940        99999  1929     11   17  45.500000   
16           33790        99999  1929      8   12  58.299999   
17           34970        99999  1929     11   13  35.599998   
18           39530        99999  1929     12    8  43.799999   
19           34970        99999  1929     11   21  46.599998   

    num_mean_temp_samples  mean_dew_point  num_mean_dew_point_samples  \
0                       4       45.500000                           4   
1                       4       46.299999                           4   
2                       4       44.700001                           4   
3                       4       41.700001                           4   
4                       4       37.700001                           4   
5                       4       43.500000                           4   
6                       4       42.299999                           4   
7                       4             NaN                        <NA>   
8                       4       48.000000                           4   
9                       4       37.700001                           4   
10                      4       40.500000                           4   
11                      4       46.700001                           4   
12                      4       46.000000                           4   
13                      4       31.000000                           4   
14                      4       40.500000                           4   
15                      4       38.500000                           4   
16                      4       50.799999                           4   
17                      5       33.200001                           4   
18                      5       39.500000                           4   
19                      5       45.299999                           4   

    mean_sealevel_pressure  ...  min_temperature  min_temperature_explicit  \
0               992.299988  ...              NaN                      <NA>   
1               987.000000  ...              NaN                      <NA>   
2              1015.299988  ...              NaN                      <NA>   
3               993.900024  ...              NaN                      <NA>   
4               991.200012  ...              NaN                      <NA>   
5              1008.700012  ...              NaN                      <NA>   
6               995.500000  ...              NaN                      <NA>   
7              1002.700012  ...              NaN                      <NA>   
8              1016.299988  ...              NaN                      <NA>   
9              1029.400024  ...              NaN                      <NA>   
10              995.400024  ...              NaN                      <NA>   
11             1006.700012  ...              NaN                      <NA>   
12              991.799988  ...              NaN                      <NA>   
13             1039.500000  ...  

## Part 1

### 1. Task
Change the date format to 'YYYY-MM-DD' and select the data from 2005 till 2009 for station numbers including and between 725300 and 726300 , and save it as a pandas dataframe. Note the maximum year available is 2010. 

#### 1. First, let us check if the maximum year is 2010.

We can do this by selecting the maximum year in the gsod data and saving it under the name "latest_year". We do the same with the minimum year.

The latest year is 2010. Good! The earliest year is 1929.

In [9]:
%%bigquery
SELECT 
    min(year) earliest_year,
    max(year) latest_year
FROM
    `bigquery-public-data.samples.gsod`

Downloading: 100%|██████████| 1/1 [00:02<00:00,  2.59s/rows]


earliest_year  latest_year
0           1929         2010

#### 2. Get the right date format and select only the relevant stations and years. 

- We can concatinate the 3 columns year, month and day into one "date" column.
- Then we select only the station numbers between 725300 AND 726300, which will also include the stations with the numbers 725300 and 726300. Between is inclusive.
- After that, we only select the years 2005 and 2009 from the data and limit the output to the top 20 rows. That is enough to get an overview of our new data set.

In [10]:
%%bigquery
SELECT
      *
    , CONCAT(year,'-',month,'-',day) as date
FROM
    `bigquery-public-data.samples.gsod`
WHERE
    station_number BETWEEN 725300 AND 726300
    AND year BETWEEN 2005 AND 2009
LIMIT 20

Downloading: 100%|██████████| 20/20 [00:01<00:00, 11.52rows/s]


station_number  wban_number  year  month  day  mean_temp  \
0           725835        99999  2005      1    2  23.000000   
1           726183        99999  2005     10   26  43.200001   
2           725940        99999  2005      5    2  52.700001   
3           725940        99999  2005     10   31  50.799999   
4           725827        99999  2005     11    7  46.000000   
5           725848        99999  2005      8   30  70.699997   
6           725940        99999  2005     10    6  51.500000   
7           725940        99999  2005      7   16  56.299999   
8           725835        99999  2005      9   22  75.599998   
9           725835        99999  2005      6   15  81.000000   
10          725469        99999  2005     11   27  52.200001   
11          725835        99999  2005     12   29  40.400002   
12          725835        99999  2005      1   27  36.200001   
13          725835        99999  2005      8    9  91.400002   
14          725868        99999  2005      3   26  34.599998   
15          725786        99999  2005      3   21  39.599998   
16          725468        99999  2005     10   31  52.500000   
17          725327        99999  2005     10   25  49.299999   
18          725976        99999  2005      5   13  46.900002   
19          725456        99999  2005     10   31  50.500000   

    num_mean_temp_samples  mean_dew_point  num_mean_dew_point_samples  \
0                       4       19.400000                           4   
1                       4       30.200001                           4   
2                       4       48.000000                           4   
3                       4       44.200001                           4   
4                       4       30.700001                           4   
5                       4       46.900002                           4   
6                       4       47.000000                           4   
7                       4       52.500000                           4   
8                       5       34.500000                           5   
9                       5       30.600000                           5   
10                      5       43.900002                           5   
11                      6       31.100000                           6   
12                      6       31.700001                           6   
13                      6       43.400002                           6   
14                      7       30.700001                           7   
15                      9       33.400002                           9   
16                     10       35.400002                          10   
17                     10       37.400002                          10   
18                     11       41.799999                          11   
19                     11       43.500000                          11   

    mean_sealevel_pressure  ...  min_temperature_explicit  \
0                      NaN  ...                      <NA>   
1                      NaN  ...                      <NA>   
2              1020.000000  ...                      <NA>   
3              1029.500000  ...                      <NA>   
4              1007.700012  ...                      <NA>   
5              1015.000000  ...                      <NA>   
6              1018.000000  ...                      <NA>   
7              1014.700012  ...                      <NA>   
8                      NaN  ...                      <NA>   
9                      NaN  ...                      <NA>   
10                     NaN  ...                      <NA>   
11                     NaN  ...                      <NA>   
12                     NaN  ...                      <NA>   
13                     NaN  ...                      <NA>   
14                     NaN  ...                      <NA>   
15                     NaN  ...                      <NA>   
16                     NaN  ...                      <NA>   
17             1017.799988 

#### 3. Save output as a pandas data frame  

The next task is to download the output of big query into a pandas data frame. For this we need:
- Pandas as pd for downloading the data into a pandas data frame.
- Service_account from google.oauth2 to save the credentials as needed.
- The query string we saw in the task before.

In [11]:
import pandas as pd
from google.oauth2 import service_account

Now we save the credentials in the right format:
 1. Location of the json file
 2. Scope is Google API

In [12]:
credentials = service_account.Credentials.from_service_account_file(
    "C:/Users/jana/Coding Challenge/learnings-code-challenges-bc3ef22c46f8.json",
    scopes=['https://www.googleapis.com/auth/cloud-platform'],
)

As the query string, we use the query we already saw before.
<br> Note: As the concat-argument for creating the date column does not work when downloading the data, I deleted this part here. I will create another date column in pandas later.

In [13]:
query_string = 'SELECT * FROM `bigquery-public-data.samples.gsod` WHERE station_number BETWEEN 725300 AND 726300 AND year BETWEEN 2005 AND 2009;'

Then we download the data and save it as a pandas data frame with the name df.

In [14]:
df = pd.read_gbq(query_string, credentials = credentials)

Downloading: 100%|██████████| 377784/377784 [02:07<00:00, 2968.47rows/s]


Now, we can check if it is a Pandas DataFrame:

In [15]:
isinstance(df, pd.DataFrame)

True

When the downloading process is completed, let us look at the first 5 rows of the df:

In [16]:
df.head()

station_number  wban_number  year  month  day  mean_temp  \
0          725835        99999  2005      1    1  28.400000   
1          725848        99999  2005      2   17  57.200001   
2          725827        99999  2005      2   24  13.100000   
3          725940        99999  2005      3   28  51.299999   
4          725940        99999  2005      8    5  53.200001   

   num_mean_temp_samples  mean_dew_point  num_mean_dew_point_samples  \
0                      4       24.299999                           4   
1                      4       46.000000                           4   
2                      4        7.700000                           4   
3                      4       45.500000                           4   
4                      4       51.299999                           4   

   mean_sealevel_pressure  ...  min_temperature  min_temperature_explicit  \
0                     NaN  ...              NaN                      <NA>   
1             1014.400024  ...              NaN                      <NA>   
2             1023.500000  ...              NaN                      <NA>   
3             1010.000000  ...              NaN                      <NA>   
4             1016.799988  ...              NaN                      <NA>   

   total_precipitation  snow_depth    fog   rain   snow   hail  thunder  \
0                 0.09         7.1  False  False  False  False    False   
1                 0.00         NaN  False  False  False  False    False   
2                 0.00        20.1  False  False  False  False    False   
3                 0.74         NaN  False  False  False  False    False   
4                  NaN         NaN  False  False  False  False    False   

   tornado  
0    False  
1    False  
2    False  
3    False  
4    False  

[5 rows x 31 columns]

Lastly, I save the df in csv format.

In [17]:
df.to_csv("data_7learnings_challenge", encoding='utf-8', index=False)

In [18]:
df = pd.read_csv("data_7learnings_challenge")  

### 2. Task 
#### From here want to work with the data from all stations 725300 to 725330 that have information from 2005 till 2009. 

First, I have a look on how large the data set is at the beginning.

In [19]:
len(df)

377784

We only need weather stations from 725300 to 725330. Lets only select the stations with the right numbers.

In [20]:
df = df.loc[(df['station_number'].between(725300, 725330, inclusive = True))]

We have only 10 stations left with numbers between 725300 and 725330.

In [21]:
df['station_number'].nunique()

10

#### Now, lets check if we need to exclude any other station, which has no data in the years between 2005 and 2009.

Not every station has data from every day, e.g., station 725314 only has 350 entrys in the year 2005. But every station has at least some data per year. We do not have to exclude any further station.

In [22]:
pd.crosstab(df['year'], df['station_number'])

station_number  725300  725305  725314  725315  725316  725317  725320  \
year                                                                     
2005               365     365     350     365     365     364     365   
2006               365     365     357     365     365     365     365   
2007               363     364     364     364     364     364     364   
2008               364     365     364     365     365     364     365   
2009               365     365     365     365     365     365     365   

station_number  725326  725327  725330  
year                                    
2005               365     358     365  
2006               359     333     365  
2007               364     364     364  
2008               361     363     364  
2009               336     363     365

#### Create date column

After that, we create the date column again, which was not downloaded via query. We create a date column that contains the information of year, month and day and saves it in the date format 'YYYY-MM-DD'.

In [23]:
df['date'] = pd.to_datetime(df['year'].map(str) + '-' + df['month'].map(str) + '-' + df['day'].map(str))

In [24]:
df.sort_values(by='date', inplace=True)

In [25]:
df.head().T

138156               276587  \
station_number                                   725317               725330   
wban_number                                       99999                14827   
year                                               2005                 2005   
month                                                 1                    1   
day                                                   1                    1   
mean_temp                                          48.6                 38.4   
num_mean_temp_samples                                24                   24   
mean_dew_point                                       46                 32.6   
num_mean_dew_point_samples                           24                   24   
mean_sealevel_pressure                           1025.4               1027.8   
num_mean_sealevel_pressure_samples                   15                   24   
mean_station_pressure                               NaN                  NaN   
num_mean_station_pressure_samples                   NaN                  NaN   
mean_visibility                                     4.8                  9.2   
num_mean_visibility_samples                          24                   24   
mean_wind_speed                                       7                  7.9   
num_mean_wind_speed_samples                          24                   24   
max_sustained_wind_speed                             13                 15.9   
max_gust_wind_speed                                 NaN                  NaN   
max_temperature                                      41                   30   
max_temperature_explicit                          False                False   
min_temperature                                     NaN                  NaN   
min_temperature_explicit                            NaN                  NaN   
total_precipitation                                0.03                 0.08   
snow_depth                                          NaN                  NaN   
fog                                                True                 True   
rain                                               True                 True   
snow                                               True                 True   
hail                                               True                 True   
thunder                                            True                 True   
tornado                                            True                 True   
date                                2005-01-01 00:00:00  2005-01-01 00:00:00   

                                                 279808               270172  \
station_number                                   725327               725305   
wban_number                                       99999                99999   
year                                               2005                 2005   
month                                                 1                    1   
day                                                   1                    1   
mean_temp                                          34.2                 30.9   
num_mean_temp_samples                                24                   24   
mean_dew_point                                     27.4                 21.8   
num_mean_dew_point_samples                           24                   24   
mean_sealevel_pressure                           1028.2               1028.7   
num_mean_sealevel_pressure_samples                   23                   19   
mean_station_pressure                               NaN                  NaN   
num_mean_station_pressure_samples                   NaN                  NaN   
mean_visibility                                     9.8                   10   
num_mean_visibility_samples                          24                   24   
mean_wind_speed                                     7.6                  8.6   
num_mean_wind_speed_samples                     

# Exploratory Data Analysis

#### Do a first analysis of the remaining dataset, clean or drop data depending on how you see appropriate. 

#### Distribution of numeric variables

We have:  
- 18.129 entrys left in our data set
- Stations between 725300 an 725330
- Years from 2005-2009
- The minimum of the average temperature is -16.20 
- and the maximum is 92

In [26]:
df.describe().round(2)

station_number  wban_number      year     month       day  mean_temp  \
count        18129.00     18129.00  18129.00  18129.00  18129.00   18129.00   
mean        725316.96     72766.47   2007.00      6.53     15.72      52.00   
std              8.98     39561.09      1.41      3.45      8.80      19.39   
min         725300.00      3887.00   2005.00      1.00      1.00     -16.20   
25%         725314.00     14842.00   2006.00      4.00      8.00      36.30   
50%         725316.00     99999.00   2007.00      7.00     16.00      53.80   
75%         725326.00     99999.00   2008.00     10.00     23.00      69.10   
max         725330.00     99999.00   2009.00     12.00     31.00      92.00   

       num_mean_temp_samples  mean_dew_point  num_mean_dew_point_samples  \
count               18129.00        18127.00                    18127.00   
mean                   23.57           41.61                       23.55   
std                     1.82           18.49                        1.86   
min                     4.00          -23.50                        4.00   
25%                    24.00           27.50                       24.00   
50%                    24.00           42.60                       24.00   
75%                    24.00           57.20                       24.00   
max                    24.00           79.10                       24.00   

       mean_sealevel_pressure  ...  num_mean_visibility_samples  \
count                16291.00  ...                     18126.00   
mean                  1016.94  ...                        23.57   
std                      6.92  ...                         1.83   
min                    985.30  ...                         4.00   
25%                   1012.80  ...                        24.00   
50%                   1016.70  ...                        24.00   
75%                   1020.90  ...                        24.00   
max                   1044.20  ...                        24.00   

       mean_wind_speed  num_mean_wind_speed_samples  max_sustained_wind_speed  \
count         18124.00                     18124.00                  18119.00   
mean              7.50                        23.53                     14.05   
std               3.58                         1.93                      5.24   
min               0.00                         4.00                      2.90   
25%               4.80                        24.00                      9.90   
50%               7.00                        24.00                     13.00   
75%               9.60                        24.00                     16.90   
max              24.00                        24.00                     49.00   

       max_gust_wind_speed  max_temperature  min_temperature  \
count             11468.00         18127.00              0.0   
mean                 23.15            42.58              NaN   
std                   6.38            18.65              NaN   
min                  11.10           -34.60              NaN   
25%                  18.10            28.40              NaN   
50%                  22.00            44.10              NaN   
75%                  27.00            57.90              NaN   
max                  63.90            87.80              NaN   

       min_temperature_explicit  total_precipitation  snow_depth  
count                       0.0             17938.00      565.00  
mean                        NaN                 0.10        2.87  
std                         NaN                 0.28        2.11  
min                         NaN                 0.00        1.20  
25%                         NaN                 0.00        1.20  
50%                         NaN                 0.00        2.00  
75%                         NaN                 0.04        3.90  
max                         NaN                 9.99       13.00  

[8 rows x 24 columns]

#### Missing values

When we look at the missing values in the data frame, we see that we have some columns with missing values. We will drop all the variables which have a lot of missing values.


Other ways could be to drop the observations with missing values instead of full columns, to impute the missing values with e. g., the mean of the existing values of the column, with 0 (only if this makes sense in terms of content e. g., sales on Sunday) or by predicting them with a model.

In [27]:
na = df.isna().sum()
na[na > 0].sort_values(ascending = False)

min_temperature_explicit              18129
min_temperature                       18129
snow_depth                            17564
num_mean_station_pressure_samples     17136
mean_station_pressure                 17136
max_gust_wind_speed                    6661
mean_sealevel_pressure                 1838
num_mean_sealevel_pressure_samples     1838
total_precipitation                     191
max_sustained_wind_speed                 10
mean_wind_speed                           5
num_mean_wind_speed_samples               5
mean_visibility                           3
num_mean_visibility_samples               3
max_temperature                           2
max_temperature_explicit                  2
num_mean_dew_point_samples                2
mean_dew_point                            2
dtype: int64

We drop all columns with more than 10% missing values.

In [28]:
na[na > len(df)*0.1].sort_values(ascending=False)

min_temperature_explicit              18129
min_temperature                       18129
snow_depth                            17564
num_mean_station_pressure_samples     17136
mean_station_pressure                 17136
max_gust_wind_speed                    6661
num_mean_sealevel_pressure_samples     1838
mean_sealevel_pressure                 1838
dtype: int64

In [29]:
df = df.drop(columns = ['mean_sealevel_pressure', 'num_mean_sealevel_pressure_samples', 'mean_station_pressure', 'num_mean_station_pressure_samples', 'max_gust_wind_speed', 'min_temperature', 'min_temperature_explicit', 'snow_depth'])

I will keep the observations that contain missing values since XGBoost can handle them.

In [30]:
na_after_droping = df.isna().sum()
na_after_droping[na_after_droping > 0].sort_values(ascending=False)

total_precipitation            191
max_sustained_wind_speed        10
num_mean_wind_speed_samples      5
mean_wind_speed                  5
num_mean_visibility_samples      3
mean_visibility                  3
max_temperature_explicit         2
max_temperature                  2
num_mean_dew_point_samples       2
mean_dew_point                   2
dtype: int64

Now, we have 18129 observations left.

In [31]:
len(df)

18129

#### Data types

After taking care of missing values, we have a look on the types of variables in our data set. We have 3 different typ of data: Integer, float and boolean.

In [32]:
df.dtypes

station_number                          int64
wban_number                             int64
year                                    int64
month                                   int64
day                                     int64
mean_temp                             float64
num_mean_temp_samples                   int64
mean_dew_point                        float64
num_mean_dew_point_samples            float64
mean_visibility                       float64
num_mean_visibility_samples           float64
mean_wind_speed                       float64
num_mean_wind_speed_samples           float64
max_sustained_wind_speed              float64
max_temperature                       float64
max_temperature_explicit               object
total_precipitation                   float64
fog                                      bool
rain                                     bool
snow                                     bool
hail                                     bool
thunder                           

Changing all boolean variables into integer to use it in the XGBoost model later.

In [33]:
df.replace({False: 0, True: 1}, inplace=True)

In [34]:
df.dtypes

station_number                          int64
wban_number                             int64
year                                    int64
month                                   int64
day                                     int64
mean_temp                             float64
num_mean_temp_samples                   int64
mean_dew_point                        float64
num_mean_dew_point_samples            float64
mean_visibility                       float64
num_mean_visibility_samples           float64
mean_wind_speed                       float64
num_mean_wind_speed_samples           float64
max_sustained_wind_speed              float64
max_temperature                       float64
max_temperature_explicit              float64
total_precipitation                   float64
fog                                     int64
rain                                    int64
snow                                    int64
hail                                    int64
thunder                           

## 3. Task
Now it is time to split the data, into a training, evaluation and test set. As a reminder, the date we are trying to predict snow fall for is the following, and hence should constitute your test set.

In [35]:
import datetime as dt

pred_date = str(dt.datetime.today() - dt.timedelta(days=13*365)).split(' ')[0]
pred_date

'2009-09-22'

In [36]:
from sklearn.model_selection import train_test_split

#### Train and Test set 

Split train and test data into 2 data frames. The date we want to predict is '2009-09-22'. This date (and every date after it) comes into the test set. All the dates before '2009-09-22' are saved in the train and validation set.

In [37]:
df_full_train = df[df['date'] < pred_date]

In [38]:
df_test = df[df['date'] >= pred_date]

Check, if we defined the test data correctly:

In [39]:
print(df_test['date'].min())
print(df_test['date'].max())

2009-09-22 00:00:00
2009-12-31 00:00:00


Now, split train and validation set into different sets. Our validation set should be 20% of the full training set.

In [40]:
df_train, df_val = train_test_split(df_full_train, test_size = 0.2, random_state = 1)

In [41]:
len(df_train), len(df_val), len(df_test)

(13697, 3425, 1007)

Looking at the count of snow days at every station, it is most likely that at station 725326 will be no snow as there was no snow recorded in the years 2005-2009.

#### Snow days at stations

In [42]:
pd.crosstab(df_train['station_number'], df_train['snow'])

snow               0    1
station_number           
725300          1231  169
725305          1215  163
725314          1210  149
725315          1252  160
725316          1248  129
725317          1224  144
725320          1124  237
725326          1338    0
725327          1142  200
725330          1097  265

#### Target variable: snow

Save the target variable into a different variable. We don't want it to be accidentally in our data set and in our model.

In [43]:
y_train = df_train.snow.values
y_val = df_val.snow.values
y_test = df_test.snow.values

And than delete it from the data sets.

In [44]:
del df_train['snow']
del df_val['snow']
del df_test['snow']

In [45]:
del df_train['date']
del df_val['date']
del df_test['date']

Let's have a last look at our data sets.

In [46]:
df_train.head()

station_number  wban_number  year  month  day  mean_temp  \
292720          725326        99999  2005      9   22  75.099998   
229080          725326        99999  2008      9    3  69.000000   
311976          725300        94846  2009      2   12  36.000000   
110188          725317        99999  2007     12   27  36.500000   
277319          725315        99999  2006     11   28  56.099998   

        num_mean_temp_samples  mean_dew_point  num_mean_dew_point_samples  \
292720                     24       62.799999                        24.0   
229080                     24       61.099998                        24.0   
311976                     24       27.700001                        24.0   
110188                     24       33.299999                        24.0   
277319                     24       46.700001                        24.0   

        mean_visibility  ...  num_mean_wind_speed_samples  \
292720             10.0  ...                         24.0   
229080              9.9  ...                         24.0   
311976             10.0  ...                         24.0   
110188              6.0  ...                         24.0   
277319             10.0  ...                         24.0   

        max_sustained_wind_speed  max_temperature  max_temperature_explicit  \
292720                       9.9        68.000000                       1.0   
229080                      12.0        62.599998                       0.0   
311976                      23.9        32.000000                       0.0   
110188                       7.0        33.799999                       1.0   
277319                      20.0        50.000000                       0.0   

        total_precipitation  fog  rain  hail  thunder  tornado  
292720                 0.00    0     0     0        0        0  
229080                 0.03    0     0     0        0        0  
311976                 0.75    0     0     0        0        0  
110188                 0.00    0     0     0        0        0  
277319                 0.00    0     0     0        0        0  

[5 rows x 22 columns]

In [47]:
df_test.head()

station_number  wban_number  year  month  day  mean_temp  \
68485           725317        53802  2009      9   22  71.199997   
269539          725300        94846  2009      9   22  67.599998   
72029           725316         3887  2009      9   22  71.199997   
157253          725314         3960  2009      9   22  71.500000   
262718          725326         4894  2009      9   22  66.099998   

        num_mean_temp_samples  mean_dew_point  num_mean_dew_point_samples  \
68485                      24       67.900002                        24.0   
269539                     24       63.799999                        24.0   
72029                      24       65.500000                        24.0   
157253                     24       64.900002                        24.0   
262718                     24       61.400002                        24.0   

        mean_visibility  ...  num_mean_wind_speed_samples  \
68485               6.1  ...                         24.0   
269539              6.0  ...                         24.0   
72029               7.7  ...                         24.0   
157253              9.1  ...                         24.0   
262718              6.0  ...                         24.0   

        max_sustained_wind_speed  max_temperature  max_temperature_explicit  \
68485                       11.1        66.900002                       0.0   
269539                      12.0        59.000000                       0.0   
72029                       11.1        66.000000                       0.0   
157253                      11.1        64.400002                       1.0   
262718                      12.0        57.200001                       1.0   

        total_precipitation  fog  rain  hail  thunder  tornado  
68485                  0.02    0     0     0        0        0  
269539                 0.00    0     0     0        0        0  
72029                  0.31    0     0     0        0        0  
157253                 0.33    0     0     0        0        0  
262718                 0.43    0     0     0        0        0  

[5 rows x 22 columns]

In [48]:
df_val.head()

station_number  wban_number  year  month  day  mean_temp  \
44008           725317        99999  2007      4   11  49.599998   
191721          725316        99999  2006      5   30  80.900002   
61720           725326         4894  2009      8   22  64.000000   
363846          725305        94892  2009      3    1  21.100000   
199640          725316        99999  2008      7   19  78.599998   

        num_mean_temp_samples  mean_dew_point  num_mean_dew_point_samples  \
44008                      18       39.200001                        18.0   
191721                     24       65.900002                        24.0   
61720                      24       51.000000                        24.0   
363846                     22        7.000000                        22.0   
199640                     24       70.699997                        24.0   

        mean_visibility  ...  num_mean_wind_speed_samples  \
44008               7.6  ...                         18.0   
191721              9.9  ...                         24.0   
61720              10.0  ...                         24.0   
363846             10.0  ...                         22.0   
199640              7.8  ...                         24.0   

        max_sustained_wind_speed  max_temperature  max_temperature_explicit  \
44008                       20.0        42.799999                       1.0   
191721                      16.9        71.599998                       1.0   
61720                       12.0        51.799999                       1.0   
363846                      16.9        17.100000                       0.0   
199640                      12.0        73.000000                       0.0   

        total_precipitation  fog  rain  hail  thunder  tornado  
44008                  0.14    0     0     0        0        0  
191721                 0.04    0     0     0        0        0  
61720                  0.09    0     0     0        0        0  
363846                 0.00    0     0     0        0        0  
199640                 0.16    0     0     0        0        0  

[5 rows x 22 columns]

#### Feature importance

As XGBoost calculates the importance of the features automatically, we will just have a quick look at the mutual information that the different variables give us for our target variable. [I used the method from ml zoomcamp.](https://www.youtube.com/watch?v=_u2YaGT6RN0&list=PL3MmuxUbc_hIhxl5Ji8t4O6lPAOpHaCLR&index=34) I will use the mutual info score from sklearn.

In [49]:
from sklearn.metrics import mutual_info_score

Select all the features from our train data.

In [50]:
df_full_train.columns

Index(['station_number', 'wban_number', 'year', 'month', 'day', 'mean_temp',
       'num_mean_temp_samples', 'mean_dew_point', 'num_mean_dew_point_samples',
       'mean_visibility', 'num_mean_visibility_samples', 'mean_wind_speed',
       'num_mean_wind_speed_samples', 'max_sustained_wind_speed',
       'max_temperature', 'max_temperature_explicit', 'total_precipitation',
       'fog', 'rain', 'snow', 'hail', 'thunder', 'tornado', 'date'],
      dtype='object')

In [51]:
num = ['station_number', 'wban_number', 'year', 'month', 'day', 'mean_temp', 'num_mean_temp_samples',
       'mean_dew_point', 'num_mean_dew_point_samples', 'mean_visibility',
       'num_mean_visibility_samples', 'mean_wind_speed',
       'num_mean_wind_speed_samples', 'max_sustained_wind_speed',
       'max_temperature', 'max_temperature_explicit', 'total_precipitation',
       'fog', 'rain', 'snow', 'hail', 'thunder', 'tornado']

And drop all observations with missing values.

In [52]:
df_full_train_NA = df_full_train.dropna()

For each variable in our Train data set, we calculate how much information we gain for the target variable "snow". <br> <br> The other weather information brings a lot of information on our target variable as well as the date variable. That it will snow is more likely in winter than in summer, depending on the location.

In [53]:
def calculate_mi(series):
    return mutual_info_score(series, df_full_train_NA.snow)

df_mi = df_full_train_NA[num].apply(calculate_mi)
df_mi.sort_values(ascending=False).to_frame(name='MI')

MI
tornado                      0.362141
thunder                      0.362141
hail                         0.362141
snow                         0.362141
rain                         0.362141
fog                          0.362141
mean_visibility              0.104079
mean_temp                    0.037668
mean_dew_point               0.031929
total_precipitation          0.028819
station_number               0.017492
month                        0.017193
year                         0.014452
max_temperature              0.013528
mean_wind_speed              0.008987
wban_number                  0.007714
max_sustained_wind_speed     0.007419
num_mean_visibility_samples  0.002832
num_mean_temp_samples        0.002509
num_mean_wind_speed_samples  0.002507
num_mean_dew_point_samples   0.002465
day                          0.001889
max_temperature_explicit     0.000004

## Part 2
If you made it up to here all by yourself, you can use your prepared dataset to train an Algorithm of your choice to forecast whether it will snow on the following date for each station in this dataset:

You are allowed to use any library you are comfortable with such as sklearn, tensorflow keras etc. 
If you did not manage to finish part one feel free to use the data provided in 'coding_challenge.csv' Note that this data does not represent a solution to Part 1. 

#### Training the XGBoost model

Packes we nee for the model

In [54]:
import xgboost as xgb
import numpy as np

Define the Classifier

In [55]:
xg_cl = xgb.XGBClassifier(objective ='binary:logistic', n_estimators = 10, seed=123)

Fit the model to the training data.

In [56]:
model = xg_cl.fit(df_train, y_train)

Predicte values for the validation set.

In [57]:
preds_val = xg_cl.predict(df_val)

Calculate the accuracy of the predictions and the actual values.

In [58]:
accuracy = float(np.sum(preds_val == y_val))/y_val.shape[0]
print("accuracy: %f" % (accuracy))

accuracy: 1.000000


Predicte values for the test set.

In [59]:
preds = xg_cl.predict(df_test)

Calculate the accuracy of the predictions and the actual values. The accuracy seems to be perfect. If I had more time, I would double-check this. A complete perfect model is highly unlikely.

In [60]:
accuracy = float(np.sum(preds==y_test))/y_test.shape[0]
print("accuracy: %f" % (accuracy))

accuracy: 1.000000


#### Now, lets look if it will snow at our stations.

In [61]:
df_test['pred'] = preds

C:\Users\jana\anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [62]:
df_test['snow'] = y_test

C:\Users\jana\anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [63]:
df_test['date'] = pd.to_datetime(df_test['year'].map(str) + '-' + df_test['month'].map(str) + '-' + df_test['day'].map(str))

C:\Users\jana\anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [64]:
df_target_date = df_test[df_test['date'] == pred_date]

In [65]:
df_target_date[['station_number', 'pred', 'snow']]

station_number  pred  snow
68485           725317     0     0
269539          725300     0     0
72029           725316     0     0
157253          725314     0     0
262718          725326     0     0
173988          725327     1     1
183954          725330     0     0
206768          725315     1     1
246461          725320     0     0
364276          725305     1     1

It will snow at 3 of 10 stations on the 22-09-2022.